From 229a20dc4779e7e3d66dcb06ec75017bebf2c184 Mon Sep 17 00:00:00 2001
From: Heyu Hu <huheyu2@huawei.com>
Date: Mon, 9 Feb 2026 23:23:57 +0800
Subject: [PATCH] urma: ipourma resource management

Offering: URMA
URMA inclusion
category: feature
bugzilla: NA

----------------------------------------------

This patch adds ipourma resource management module
with resource initialization/deinitialization, reserved data
plane interfaces and network device options,introduces
netlink and ethtool modules.

Signed-off-by: Heyu Hu <huheyu2@huawei.com>
---
 drivers/ub/urma/ulp/ipourma/Makefile          |   3 +
 drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c |  22 +
 drivers/ub/urma/ulp/ipourma/ipourma_ethtool.h |  15 +
 drivers/ub/urma/ulp/ipourma/ipourma_netdev.c  | 105 +++-
 drivers/ub/urma/ulp/ipourma/ipourma_netlink.c |  20 +
 drivers/ub/urma/ulp/ipourma/ipourma_netlink.h |  15 +
 drivers/ub/urma/ulp/ipourma/ipourma_res.c     | 292 +++++++++-
 drivers/ub/urma/ulp/ipourma/ipourma_res.h     |   3 +
 drivers/ub/urma/ulp/ipourma/ipourma_types.h   |   1 +
 drivers/ub/urma/ulp/ipourma/ipourma_ub.c      | 501 +++++++++++++++++-
 drivers/ub/urma/ulp/ipourma/ipourma_ub.h      |   1 +
 drivers/ub/urma/ulp/ipourma/ipourma_utils.c   |  33 ++
 drivers/ub/urma/ulp/ipourma/ipourma_utils.h   |  15 +
 13 files changed, 989 insertions(+), 37 deletions(-)
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_ethtool.h
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_netlink.c
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_netlink.h
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_utils.c
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_utils.h

diff --git a/drivers/ub/urma/ulp/ipourma/Makefile b/drivers/ub/urma/ulp/ipourma/Makefile
index f50cf925b179..870fbd99b9b6 100644
--- a/drivers/ub/urma/ulp/ipourma/Makefile
+++ b/drivers/ub/urma/ulp/ipourma/Makefile
@@ -11,5 +11,8 @@ ipourma_ub.o \
 ipourma_err.o \
 ipourma_addr_res.o \
 ipourma_ip.o \
+ipourma_ethtool.o \
+ipourma_netlink.o \
+ipourma_utils.o \
 
 obj-$(CONFIG_UB_URMA) += ipourma.o
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c b/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c
new file mode 100644
index 000000000000..1943ad56c5b6
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
+ *
+ * Description: ipourma ethtool support
+ */
+
+#include <linux/kernel.h>
+#include <linux/ethtool.h>
+#include <linux/netdevice.h>
+#include <linux/version.h>
+
+#include "ub/urma/ubcore_uapi.h"
+#include "ipourma_ethtool.h"
+
+#define IPOURMA_DRIVER_VERSION	"0.1.0"
+#define IPOURMA_FW_VERSION	"0.1.0"
+
+void ipourma_set_ethtool_ops(struct net_device *dev)
+{
+	// this function will be filled in the next commit
+}
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.h b/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.h
new file mode 100644
index 000000000000..5a22a29ea217
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.h
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2021-2025. All rights reserved.
+ *
+ * Description: Functions definition of ipourma_ethtool
+ */
+
+#ifndef _IPOURMA_ETHTOOL_H
+#define _IPOURMA_ETHTOOL_H
+
+#include "ipourma_types.h"
+
+void ipourma_set_ethtool_ops(struct net_device *dev);
+
+#endif
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c b/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c
index b1f2f63adbdf..9500c8a85ee5 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c
@@ -25,6 +25,8 @@
 #include "ipourma_main.h"
 #include "ipourma_ip.h"
 #include "ipourma_addr_res.h"
+#include "ipourma_ethtool.h"
+#include "ipourma_netlink.h"
 #include "ub/urma/ubcore_uapi.h"
 #include "ipourma_netdev.h"
 
@@ -157,13 +159,108 @@ STATIC void ipourma_add_route_entry(struct work_struct *work)
 		netdev_err(priv->dev, "Failed to new route\n");
 }
 
+STATIC int ipourma_ndo_init(struct net_device *dev)
+{
+	// this function will be filled in the next commit
+	return 0;
+}
+
+STATIC void ipourma_ndo_uninit(struct net_device *dev)
+{
+	// do uninit in ipourma_stop
+}
+
+STATIC int ipourma_open(struct net_device *dev)
+{
+	// this function will be filled in the next commit
+	return 0;
+}
+
+STATIC inline void ipourma_napi_disable(struct net_device *dev)
+{
+	// this function will be filled in the next commit
+}
+
+STATIC int ipourma_stop(struct net_device *dev)
+{
+	// this function will be filled in the next commit
+	return 0;
+}
+
+STATIC int ipourma_change_mtu(struct net_device *dev, int mtu)
+{
+	// this function will be filled in the next commit
+	return 0;
+}
+
+STATIC netdev_features_t ipourma_fix_features(struct net_device *dev,
+	netdev_features_t features)
+{
+	// this function will be filled in the next commit
+	return features;
+}
+
+STATIC netdev_tx_t ipourma_start_xmit(struct sk_buff *skb, struct net_device *dev)
+{
+	// this function will be filled in the next commit
+	return NETDEV_TX_OK;
+}
+
+STATIC void ipourma_timeout(struct net_device *dev, unsigned int txqueue)
+{
+	// this function will be filled in the next commit
+}
+
+/*
+ * for real net device, return 0.
+ */
+STATIC int ipourma_get_iflink(const struct net_device *dev)
+{
+	return 0;
+}
+
+STATIC int ipourma_set_mac(struct net_device *dev, void *addr)
+{
+	// this function will be filled in the next commit
+	return 0;
+}
+
+STATIC void ipourma_get_stats(struct net_device *dev, struct rtnl_link_stats64 *stats)
+{
+	// fill in stats
+	if (IS_ERR_OR_NULL(dev) || IS_ERR_OR_NULL(stats))
+		return;
+	netdev_stats_to_stats64(stats, &dev->stats);
+}
+
+STATIC int ipourma_ioctl(struct net_device *dev, struct ifreq *ifr, int cmd)
+{
+	/* if ipourma has some private functions in the future, ioctl will call them here. */
+	return 0;
+}
+
+STATIC const struct net_device_ops ipourma_netdev_ops = {
+	.ndo_init		 = ipourma_ndo_init,
+	.ndo_uninit		 = ipourma_ndo_uninit,
+	.ndo_open		 = ipourma_open,
+	.ndo_stop		 = ipourma_stop,
+	.ndo_change_mtu		 = ipourma_change_mtu,
+	.ndo_fix_features	 = ipourma_fix_features,
+	.ndo_start_xmit		 = ipourma_start_xmit,
+	.ndo_tx_timeout		 = ipourma_timeout,
+	.ndo_get_iflink		 = ipourma_get_iflink,
+	.ndo_set_mac_address	 = ipourma_set_mac,
+	.ndo_get_stats64	 = ipourma_get_stats,
+	.ndo_eth_ioctl		 = ipourma_ioctl,
+};
+
 STATIC void ipourma_netdev_setup(struct net_device *dev)
 {
-	// dev->netdev_ops        = &ipourma_netdev_ops;
+	dev->netdev_ops        = &ipourma_netdev_ops;
+
+	ipourma_set_ethtool_ops(dev);
+	ipourma_set_rtnl_link_ops(dev);
 
-	/* ipourma_set_ethtool_ops(dev);
-	 * ipourma_set_rtnl_link_ops(dev);
-	 */
 	dev->type              = ARPHRD_ETHER;
 	dev->hard_header_len   = IPOURMA_HARD_LEN;
 	dev->addr_len          = IPOURMA_ALEN;
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_netlink.c b/drivers/ub/urma/ulp/ipourma/ipourma_netlink.c
new file mode 100644
index 000000000000..9de717540a06
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_netlink.c
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
+ *
+ * Description: ipourma netlink support
+ */
+
+#include <net/rtnetlink.h>
+#include <linux/if_arp.h>
+#include <linux/netdevice.h>
+#include <linux/module.h>
+#include <linux/netlink.h>
+#include <linux/if_link.h>
+
+#include "ipourma_netlink.h"
+
+void ipourma_set_rtnl_link_ops(struct net_device *dev)
+{
+	// this function will be filled in the next commit
+}
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_netlink.h b/drivers/ub/urma/ulp/ipourma/ipourma_netlink.h
new file mode 100644
index 000000000000..c1c59635baf5
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_netlink.h
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2021-2025. All rights reserved.
+ *
+ * Description: Functions definition of ipourma_netlink
+ */
+
+#ifndef _IPOURMA_NETLINK_H
+#define _IPOURMA_NETLINK_H
+
+#include "ipourma_types.h"
+
+void ipourma_set_rtnl_link_ops(struct net_device *dev);
+
+#endif
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_res.c b/drivers/ub/urma/ulp/ipourma/ipourma_res.c
index f8838c7f1d0a..491e499c649e 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_res.c
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_res.c
@@ -76,30 +76,60 @@ void ipourma_uninit_rx_bufs(struct ipourma_dev_priv *priv, u32 eid_idx)
 
 void ipourma_uninit_rings_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx)
 {
-	/* this function will be filled in the next commit */
+	if (!IS_ERR_OR_NULL(priv->tx_ring) && !IS_ERR_OR_NULL(priv->tx_ring[eid_idx])) {
+		ipourma_uninit_tx_bufs(priv, eid_idx);
+		vfree(priv->tx_ring[eid_idx]);
+		priv->tx_ring[eid_idx] = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->rx_ring) && !IS_ERR_OR_NULL(priv->rx_ring[eid_idx])) {
+		ipourma_uninit_rx_bufs(priv, eid_idx);
+		kfree(priv->rx_ring[eid_idx]);
+		priv->rx_ring[eid_idx] = NULL;
+	}
 }
 
-int ipourma_init_tjetty_hmap(struct net_device *dev)
+void ipourma_uninit_rings(struct net_device *dev)
 {
 	struct ipourma_dev_priv *priv = netdev_priv(dev);
-	struct ipourma_tjetty_hmap *tjetty_hmap = &priv->tjetty_lru.tjetty_hmap;
-	int ret = IPOURMA_OK;
 
-	tjetty_hmap->hash_seed = get_random_u32();
-	tjetty_hmap->buckets =
-		kcalloc(IPOURMA_TJETTY_HMAP_SIZE, sizeof(struct hlist_head), GFP_KERNEL);
-	if (IS_ERR_OR_NULL(tjetty_hmap->buckets)) {
-		netdev_warn(priv->dev, "%s\n",
-			ipourma_err_desc(IPOURMA_INIT_TJETTY_HMAP_FAILED));
-		return IPOURMA_INIT_TJETTY_HMAP_FAILED;
+	if (!IS_ERR_OR_NULL(priv->tx_ring)) {
+		for (u32 i = 0; i < IPOURMA_MAX_EID_CNT; i++) {
+			if (IS_ERR_OR_NULL(priv->tx_ring[i]))
+				continue;
+			ipourma_uninit_tx_bufs(priv, i);
+			vfree(priv->tx_ring[i]);
+			priv->tx_ring[i] = NULL;
+		}
+		kfree(priv->tx_ring);
+		priv->tx_ring = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->tx_head)) {
+		kfree(priv->tx_head);
+		priv->tx_head = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->tx_tail)) {
+		kfree(priv->tx_tail);
+		priv->tx_tail = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->tx_count)) {
+		kfree(priv->tx_count);
+		priv->tx_count = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->rx_ring)) {
+		for (u32 i = 0; i < IPOURMA_MAX_EID_CNT; i++) {
+			if (IS_ERR_OR_NULL(priv->rx_ring[i]))
+				continue;
+			ipourma_uninit_rx_bufs(priv, i);
+			kfree(priv->rx_ring[i]);
+			priv->rx_ring[i] = NULL;
+		}
+		kfree(priv->rx_ring);
+		priv->rx_ring = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->tx_ring_locks)) {
+		kfree(priv->tx_ring_locks);
+		priv->tx_ring_locks = NULL;
 	}
-
-	for (int i = 0; i < IPOURMA_TJETTY_HMAP_SIZE; i++)
-		INIT_HLIST_HEAD(tjetty_hmap->buckets + i);
-
-	ipourma_init_tjetty_aging_work(&priv->tjetty_lru);
-
-	return ret;
 }
 
 STATIC int ipourma_alloc_tx_buf_aligned(struct ipourma_dev_priv *priv, u32 eid_idx, u32 idx)
@@ -397,11 +427,34 @@ STATIC void ipourma_reset_rx_buf_by_eid(struct ipourma_dev_priv *priv,
 	}
 }
 
+STATIC inline void ipourma_unregister_rx_seg_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx)
+{
+	if (IS_ERR_OR_NULL(priv->rx_ring[eid_idx]))
+		return;
+	for (u32 i = 0; i < ipourma_rx_ring_size; i++) {
+		if (!IS_ERR_OR_NULL(priv->rx_ring[eid_idx][i].seg[0]))
+			priv->rx_ring[eid_idx][i].seg[0] = NULL;
+	}
+}
+
 STATIC int ipourma_reset_rx_bufs(struct ipourma_dev_priv *priv, struct ubcore_cr *cr)
 {
-	/* this function will be filled in the next commit */
-	ipourma_reset_rx_buf_by_eid(priv, cr, 0);
-	return 0;
+	if (IS_ERR_OR_NULL(priv->jfr))
+		return IPOURMA_OK;
+
+	for (u32 i = 0; i < IPOURMA_MAX_EID_CNT; i++) {
+		ipourma_reset_rx_buf_by_eid(priv, cr, i);
+		ipourma_unregister_rx_seg_by_eid(priv, i);
+		if (IS_ERR_OR_NULL(priv->ipourma_ub_rx_seg[i]))
+			continue;
+		for (u32 j = 0; j < priv->rx_buf_num; j++) {
+			if (IS_ERR_OR_NULL(priv->ipourma_ub_rx_seg[i][j]))
+				continue;
+			ubcore_unregister_seg(priv->ipourma_ub_rx_seg[i][j]);
+			priv->ipourma_ub_rx_seg[i][j] = NULL;
+		}
+	}
+	return IPOURMA_OK;
 }
 
 int ipourma_reset_rings(struct ipourma_dev_priv *priv)
@@ -424,13 +477,204 @@ int ipourma_reset_rings(struct ipourma_dev_priv *priv)
 	return IPOURMA_OK;
 }
 
+void ipourma_uninit_tjetty_hmap(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	cancel_delayed_work_sync(&priv->tjetty_lru.tjetty_aging_work);
+	ipourma_lru_clear(&priv->tjetty_lru);
+	kfree(priv->tjetty_lru.tjetty_hmap.buckets);
+	priv->tjetty_lru.tjetty_hmap.buckets = NULL;
+}
+
+int ipourma_init_tjetty_hmap(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	struct ipourma_tjetty_hmap *tjetty_hmap = &priv->tjetty_lru.tjetty_hmap;
+	int ret = IPOURMA_OK;
+
+	tjetty_hmap->hash_seed = get_random_u32();
+	tjetty_hmap->buckets =
+		kcalloc(IPOURMA_TJETTY_HMAP_SIZE, sizeof(struct hlist_head), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(tjetty_hmap->buckets)) {
+		netdev_warn(priv->dev, "%s\n",
+			ipourma_err_desc(IPOURMA_INIT_TJETTY_HMAP_FAILED));
+		return IPOURMA_INIT_TJETTY_HMAP_FAILED;
+	}
+
+	for (int i = 0; i < IPOURMA_TJETTY_HMAP_SIZE; i++)
+		INIT_HLIST_HEAD(tjetty_hmap->buckets + i);
+
+	ipourma_init_tjetty_aging_work(&priv->tjetty_lru);
+
+	return ret;
+}
+
+STATIC struct ubcore_jfr *ipourma_create_jfr(
+	struct net_device *dev, u32 depth, u32 eid_index)
+{
+	struct ubcore_jfr_cfg jfr_cfg = { 0 };
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	jfr_cfg.depth = depth;
+	jfr_cfg.flag.bs.token_policy = UBCORE_TOKEN_NONE;
+	jfr_cfg.trans_mode = UBCORE_TP_RM;
+	jfr_cfg.eid_index = priv->eid_info[eid_index].eid_index;
+	jfr_cfg.max_sge = IPOURMA_MAX_URMA_RECV_SGES;
+	jfr_cfg.jfc = priv->rx_jfc;
+
+	return ubcore_create_jfr(priv->urma_dev, &jfr_cfg, NULL, NULL);
+}
+
+STATIC struct ubcore_jetty *ipourma_create_jetty(struct net_device *dev,
+	u32 jetty_id, u32 eid_index)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	struct ubcore_jetty_cfg jetty_cfg = {0};
+	int max_send_sge;
+	int max_recv_sge;
+
+	max_send_sge = (IPOURMA_MAX_TX_SGES < IPOURMA_MAX_URMA_SEND_SGES) ?
+		IPOURMA_MAX_TX_SGES : IPOURMA_MAX_URMA_SEND_SGES;
+	max_recv_sge = (IPOURMA_MAX_RX_SGES < IPOURMA_MAX_URMA_RECV_SGES) ?
+		IPOURMA_MAX_RX_SGES : IPOURMA_MAX_URMA_RECV_SGES;
+
+	/* some values should dynamically get from the device */
+	jetty_cfg.id = jetty_id;
+	jetty_cfg.flag.bs.share_jfr = 1;
+	jetty_cfg.trans_mode = UBCORE_TP_RM;
+	jetty_cfg.eid_index = priv->eid_info[eid_index].eid_index;
+	jetty_cfg.jfs_depth = IPOURMA_JFS_DEPTH;
+	/* jetty_cfg.priority = 0; */
+	jetty_cfg.max_send_sge = max_send_sge;
+	jetty_cfg.max_send_rsge = IPOURMA_MAX_URMA_RECV_SGES;
+	jetty_cfg.jfr_depth = IPOURMA_JFR_DEPTH;
+	jetty_cfg.max_recv_sge = max_recv_sge;
+	jetty_cfg.send_jfc = priv->tx_jfc;
+	jetty_cfg.recv_jfc = priv->rx_jfc;
+	jetty_cfg.jfr = priv->jfr[eid_index];
+	/* if (ipourma_jetty_set_priority(priv, &jetty_cfg) != IPOURMA_OK)
+	 *	return NULL;
+	 */
+
+	return ubcore_create_jetty(priv->urma_dev, &jetty_cfg, NULL, NULL);
+}
+
+STATIC void ipourma_uninit_misc(struct ipourma_dev_priv *priv)
+{
+	if (!IS_ERR_OR_NULL(priv->tjetty_lru.tjetty_wq)) {
+		flush_workqueue(priv->tjetty_lru.tjetty_wq);
+		destroy_workqueue(priv->tjetty_lru.tjetty_wq);
+		priv->tjetty_lru.tjetty_wq = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->net_config_wq)) {
+		flush_workqueue(priv->net_config_wq);
+		destroy_workqueue(priv->net_config_wq);
+		priv->net_config_wq = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->rx_wq)) {
+		destroy_workqueue(priv->rx_wq);
+		priv->rx_wq = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->tx_wq)) {
+		destroy_workqueue(priv->tx_wq);
+		priv->tx_wq = NULL;
+	}
+}
+
+STATIC void ipourma_uninit_urma_resources_table(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	if (!IS_ERR_OR_NULL(priv->jetty)) {
+		kfree(priv->jetty);
+		priv->jetty = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->jfr)) {
+		kfree(priv->jfr);
+		priv->jfr = NULL;
+	}
+}
+
 void ipourma_uninit_urma_resources_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx)
 {
-	/* this function will be filled in the next commit */
+	if (!IS_ERR_OR_NULL(priv->jetty) && !IS_ERR_OR_NULL(priv->jetty[eid_idx])) {
+		ubcore_delete_jetty(priv->jetty[eid_idx]);
+		priv->jetty[eid_idx] = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->jfr) && !IS_ERR_OR_NULL(priv->jfr[eid_idx])) {
+		ubcore_delete_jfr(priv->jfr[eid_idx]);
+		priv->jfr[eid_idx] = NULL;
+	}
+}
+
+void ipourma_uninit_urma_resources(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	ipourma_uninit_misc(priv);
+
+	if (IS_ERR_OR_NULL(priv->jetty))
+		return;
+	for (int i = 0; i < IPOURMA_MAX_EID_CNT; i++)
+		ipourma_uninit_urma_resources_by_eid(priv, i);
+	ipourma_uninit_urma_resources_table(dev);
+	if (!IS_ERR_OR_NULL(priv->rx_jfc)) {
+		ubcore_delete_jfc(priv->rx_jfc);
+		priv->rx_jfc = NULL;
+	}
+	if (!IS_ERR_OR_NULL(priv->tx_jfc)) {
+		ubcore_delete_jfc(priv->tx_jfc);
+		priv->tx_jfc = NULL;
+	}
+}
+
+static bool ipourma_check_dev_name(struct net_device *dev)
+{
+	char buf[IPOURMA_MAX_DEV_NAME] = {0};
+	u16 dev_num = 0;
+	int ret = 0;
+
+	if (IS_ERR_OR_NULL(dev) || IS_ERR_OR_NULL(dev->name))
+		return false;
+	ret = sscanf(dev->name, "ipourma%hu", &dev_num);
+	if (ret != 1)
+		return false;
+	ret = snprintf(buf, IPOURMA_MAX_DEV_NAME, "ipourma%hu", dev_num);
+	if (ret <= 0 || strcmp(buf, dev->name) != 0)
+		return false;
+	return true;
 }
 
 int ipourma_init_urma_resources_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx)
 {
-	/* this function will be filled in the next commit */
-	return 0;
+	struct net_device *dev = priv->dev;
+	int ret = IPOURMA_OK;
+
+	if (!ipourma_check_dev_name(dev))
+		goto invalid_name;
+
+	priv->jfr[eid_idx] = ipourma_create_jfr(dev, IPOURMA_JFR_DEPTH, eid_idx);
+	if (IS_ERR_OR_NULL(priv->jfr[eid_idx])) {
+		ret = IPOURMA_CREATE_JFR_FAILED;
+		pr_err("create jfr error, dev: %s, i = %d\n", dev->name, eid_idx);
+		goto jfr_failed;
+	}
+	priv->jetty[eid_idx] = ipourma_create_jetty(dev,
+				IPOURMA_WELL_KNOWN_JETTY_ID + eid_idx, eid_idx);
+	if (IS_ERR_OR_NULL(priv->jetty[eid_idx])) {
+		ret = IPOURMA_CREATE_JETTY_FAILED;
+		pr_err("create tx jetty error, dev: %s, i = %d\n", dev->name, eid_idx);
+		goto jetty_failed;
+	}
+
+	return ret;
+
+jetty_failed:
+	ubcore_delete_jfr(priv->jfr[eid_idx]);
+	priv->jfr[eid_idx] = NULL;
+jfr_failed:
+	return ret;
+invalid_name:
+	return ret;
 }
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_res.h b/drivers/ub/urma/ulp/ipourma/ipourma_res.h
index be262a630ced..25fe4e72bc73 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_res.h
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_res.h
@@ -12,7 +12,10 @@
 
 void ipourma_uninit_rings_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
 void ipourma_uninit_rx_bufs(struct ipourma_dev_priv *priv, u32 eid_idx);
+void ipourma_uninit_rings(struct net_device *dev);
 int ipourma_init_tjetty_hmap(struct net_device *dev);
+void ipourma_uninit_tjetty_hmap(struct net_device *dev);
+void ipourma_uninit_urma_resources(struct net_device *dev);
 int ipourma_reset_rings(struct ipourma_dev_priv *priv);
 int ipourma_init_rings_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
 void ipourma_uninit_urma_resources_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_types.h b/drivers/ub/urma/ulp/ipourma/ipourma_types.h
index af26a7bf86cc..38138e6e9b09 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_types.h
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_types.h
@@ -276,6 +276,7 @@ struct ipourma_tjetty_lru {
 	uint16_t count;
 	uint16_t tjetty_capacity;
 	struct delayed_work tjetty_aging_work;
+	struct workqueue_struct *tjetty_wq;
 	uint16_t tjetty_aging_interval_s;
 	uint16_t tjetty_aging_timeout_s;
 };
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_ub.c b/drivers/ub/urma/ulp/ipourma/ipourma_ub.c
index fc9f152f0e2c..496fa701c1f7 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_ub.c
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_ub.c
@@ -12,6 +12,8 @@
 #include <linux/jiffies.h>
 #include "ipourma_res.h"
 #include "ipourma_main.h"
+#include "ipourma_utils.h"
+#include "ipourma_err.h"
 #include "ub/urma/ubcore_uapi.h"
 #include "ipourma_ub.h"
 
@@ -28,30 +30,355 @@ void ipourma_build_seg_cfg(struct ubcore_seg_cfg *cfg,
 	cfg->flag = flag;
 }
 
+STATIC int ipourma_prepare_tx_data(struct ubcore_device *urma_dev,
+					struct ipourma_dev_priv *priv,
+					struct ipourma_tx_buf *tx_req)
+{
+	struct sk_buff *skb = tx_req->skb;
+	struct ubcore_seg_cfg cfg = {0};
+	u32 offset = 1, i, j;
+	u32 linear_len = skb_headlen(skb);
+
+	pr_skb_head_plus_linear(skb, "Register tx seg");
+	if (unlikely(linear_len > priv->tx_buf_size)) {
+		priv->runtime_stats.tx_stats.linear_len_oversize++;
+		netdev_dbg(priv->dev, "%s: linear len %u, tx buffer size %u\n",
+					ipourma_err_desc(IPOURMA_UNSUPPORTED_LINEAR_DATA_LEN),
+					linear_len, priv->tx_buf_size);
+		return IPOURMA_UNSUPPORTED_LINEAR_DATA_LEN;
+	}
+	if (likely(linear_len > 0))
+		memcpy(tx_req->buf_aligned, skb->data, linear_len);
+	for (i = 0; i < skb_shinfo(skb)->nr_frags; i++) {
+		skb_frag_t *frag = &skb_shinfo(skb)->frags[i];
+		void *vaddr = skb_frag_address(frag);
+		unsigned int frag_len = skb_frag_size(frag);
+
+		ipourma_build_seg_cfg(&cfg, (u64)vaddr, frag_len);
+		tx_req->seg[i + offset] = ubcore_register_seg(urma_dev, &cfg, NULL);
+		if (IS_ERR_OR_NULL(tx_req->seg[i + offset]))
+			goto partial_reg_out;
+	}
+	return IPOURMA_OK;
+
+partial_reg_out:
+	for (j = 1; j < i + offset; j++) {
+		if (IS_ERR_OR_NULL(tx_req->seg[j]))
+			continue;
+		ubcore_unregister_seg(tx_req->seg[j]);
+	}
+	return IPOURMA_REGISTER_SEG_FAILED;
+}
+
+STATIC int ipourma_setup_sges(struct ipourma_dev_priv *priv,
+	struct ipourma_tx_buf *tx_req)
+{
+	int i, offset = 0;
+	struct sk_buff *skb = tx_req->skb;
+	skb_frag_t *frags = skb_shinfo(skb)->frags;
+	int nr_frags = skb_shinfo(skb)->nr_frags;
+
+	if (skb_headlen(skb)) {
+		tx_req->tx_sge[0].len = skb_headlen(skb);
+		offset++;
+	}
+	for (i = 0; i < nr_frags; i++) {
+		tx_req->tx_sge[i + offset].addr = tx_req->seg[i + offset]->seg.ubva.va;
+		tx_req->tx_sge[i + offset].len = skb_frag_size(&frags[i]);
+		tx_req->tx_sge[i + offset].tseg = tx_req->seg[i + offset];
+	}
+
+	return (nr_frags + offset);
+}
+
+STATIC struct ipourma_tjetty_hash_node *ipourma_locate_tjetty_node(
+					struct ipourma_tjetty_lru *tjetty_lru,
+					union ubcore_eid *src_eid,
+					union ubcore_eid *dst_eid,
+					bool lock_free)
+{
+	// this function will be filled in the next commit
+	return NULL;
+}
+
+STATIC inline void ipourma_delete_tjetty_node(struct ipourma_tjetty_hash_node *tjetty_node)
+{
+	hlist_del(&tjetty_node->hlist);
+	ubcore_unimport_jetty(tjetty_node->tjetty);
+	kfree(tjetty_node);
+}
+
+STATIC void ipourma_lru_del_tail(struct ipourma_tjetty_lru *tjetty_lru, bool lock_free)
+{
+	struct ipourma_tjetty_hash_node *tjetty_node = NULL;
+
+	if (!lock_free)
+		spin_lock(&tjetty_lru->lock);
+
+	if (likely(!list_empty(&tjetty_lru->list))) {
+		tjetty_node = list_last_entry(&tjetty_lru->list,
+						struct ipourma_tjetty_hash_node,
+						lru_list);
+		list_del(&tjetty_node->lru_list);
+		ipourma_delete_tjetty_node(tjetty_node);
+		tjetty_lru->count--;
+	}
+
+	if (!lock_free)
+		spin_unlock(&tjetty_lru->lock);
+}
+
+void ipourma_lru_clear(struct ipourma_tjetty_lru *tjetty_lru)
+{
+	spin_lock(&tjetty_lru->lock);
+	while (tjetty_lru->count > 0)
+		ipourma_lru_del_tail(tjetty_lru, true);
+	spin_unlock(&tjetty_lru->lock);
+}
+
+STATIC void tjetty_aging_callback(struct work_struct *work)
+{
+	struct delayed_work *tjetty_aging_work = container_of(work,
+							      struct delayed_work,
+							      work);
+	struct ipourma_tjetty_lru *tjetty_lru = container_of(tjetty_aging_work,
+							     struct ipourma_tjetty_lru,
+							     tjetty_aging_work);
+	struct ipourma_tjetty_hash_node *tjetty_node = NULL;
+	u64 now_jiffies = 0;
+
+	if (!spin_trylock(&tjetty_lru->lock)) {
+		schedule_delayed_work(tjetty_aging_work,
+			msecs_to_jiffies(IPOURMA_TJETTY_CB_S * MSEC_PER_SEC));
+		return;
+	}
+	now_jiffies = get_jiffies_64();
+	while (tjetty_lru->count > 0) {
+		tjetty_node = list_last_entry(&tjetty_lru->list,
+					struct ipourma_tjetty_hash_node,
+					lru_list);
+		if (now_jiffies - tjetty_node->last_jiffies
+			< (u64)tjetty_lru->tjetty_aging_timeout_s * HZ)
+			break;
+		list_del(&tjetty_node->lru_list);
+		ipourma_delete_tjetty_node(tjetty_node);
+		tjetty_lru->count--;
+	}
+	schedule_delayed_work(tjetty_aging_work,
+		msecs_to_jiffies((u32)tjetty_lru->tjetty_aging_interval_s * MSEC_PER_SEC));
+	spin_unlock(&tjetty_lru->lock);
+}
+
 void ipourma_init_tjetty_aging_work(struct ipourma_tjetty_lru *tjetty_lru)
 {
-	/* this function will be filled in the next commit */
+	struct delayed_work *work = &tjetty_lru->tjetty_aging_work;
+	unsigned long flags;
+
+	spin_lock_irqsave(&tjetty_lru->lock, flags);
+	tjetty_lru->tjetty_aging_interval_s = IPOURMA_TJETTY_CB_S;
+	tjetty_lru->tjetty_aging_timeout_s = IPOURMA_TJETTY_TIMEOUT_S;
+	INIT_DELAYED_WORK(work, tjetty_aging_callback);
+	schedule_delayed_work(work,
+		msecs_to_jiffies((u32)tjetty_lru->tjetty_aging_interval_s * MSEC_PER_SEC));
+	spin_unlock_irqrestore(&tjetty_lru->lock, flags);
+}
+
+STATIC struct ubcore_tjetty *ipourma_import_new_tjetty(
+	struct ipourma_dev_priv *priv, struct ipourma_tx_buf *tx_req)
+{
+	// this function will be filled in the next commit
+	return NULL;
+}
+
+STATIC void ipourma_advance_tx_tail(struct ipourma_dev_priv *priv,
+	struct ipourma_tx_buf *tx_req)
+{
+	u32 i = 0, eid_idx = tx_req->eid_index;
+	unsigned long flags;
+
+	spin_lock_irqsave(&priv->tx_ring_locks[eid_idx], flags);
+	for (i = priv->tx_tail[eid_idx] % ipourma_tx_ring_size;
+		i != (tx_req->idx + 1) % ipourma_tx_ring_size;
+		i = (i + 1) % ipourma_tx_ring_size) {
+		if (priv->tx_ring[eid_idx][i].tx_buf_in_use == 1)
+			break;
+		priv->tx_tail[eid_idx]++;
+	}
+	if (unlikely(priv->tx_ring_is_full[eid_idx] &&
+		test_bit(IPOURMA_DEV_ADMIN_UP, &priv->flags)) &&
+		priv->tx_head[eid_idx] - priv->tx_tail[eid_idx] <= (ipourma_tx_ring_size >> 1)) {
+		priv->tx_ring_is_full[eid_idx] = false;
+		atomic_sub(1, &priv->tx_ring_blocked);
+		if (atomic_read(&priv->tx_ring_blocked) == 0)
+			netif_wake_queue(priv->dev);
+	}
+	spin_unlock_irqrestore(&priv->tx_ring_locks[eid_idx], flags);
+}
+
+STATIC inline void ipourma_handle_tx_failed(struct net_device *dev, struct ipourma_tx_buf *tx_req)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	tx_req->tx_buf_in_use = 0;
+	ipourma_advance_tx_tail(priv, tx_req);
+	dev_kfree_skb_any(tx_req->skb);
+}
+
+STATIC int ipourma_update_wr(struct net_device *dev, struct ipourma_tx_buf *tx_req)
+{
+	struct ipourma_tjetty_hash_node *tjetty_node = NULL;
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	struct ubcore_tjetty *tjetty_new = NULL;
+	union ubcore_eid *src_eid, *dst_eid;
+
+	if (IS_ERR_OR_NULL(tx_req->tjetty)) {
+		src_eid = &priv->eid_info[tx_req->eid_index].eid;
+		dst_eid = &tx_req->dst_eid;
+		tjetty_node = ipourma_locate_tjetty_node(&priv->tjetty_lru,
+								src_eid, dst_eid, false);
+		if (IS_ERR_OR_NULL(tjetty_node)) {
+			priv->runtime_stats.tx_stats.num_import_jetty_real++;
+			tjetty_new = ipourma_import_new_tjetty(priv, tx_req);
+			if (IS_ERR_OR_NULL(tjetty_new)) {
+				ipourma_handle_tx_failed(dev, tx_req);
+				priv->runtime_stats.tx_stats.import_jetty_failed++;
+				return IPOURMA_TJETTY_NODE_ALLOC_FAILED;
+			}
+		} else {
+			tjetty_new = tjetty_node->tjetty;
+			priv->runtime_stats.tx_stats.num_import_jetty_bypass++;
+		}
+		tx_req->tx_wr.tjetty = tjetty_new;
+	} else {
+		tx_req->tx_wr.tjetty = tx_req->tjetty;
+		priv->runtime_stats.tx_stats.num_tjetty_hash_hit++;
+	}
+
+	tx_req->tx_wr.send.src.num_sge = (u32)ipourma_setup_sges(priv, tx_req);
+	return 0;
 }
 
 void ipourma_post_send(struct work_struct *work)
 {
-	/* this function will be filled in the next commit */
+	struct ipourma_tx_buf *tx_req = container_of(work, struct ipourma_tx_buf, work);
+	struct ipourma_dev_priv *priv = tx_req->priv;
+	int ret;
+	struct ubcore_jfs_wr *jfs_bad_wr = NULL;
+
+	priv->runtime_stats.tx_stats.post_send_start++;
+	pr_debug("post_send start, idx %u, jetty %u\n", tx_req->idx, tx_req->eid_index);
+	ret = ipourma_prepare_tx_data(priv->urma_dev, priv, tx_req);
+	if (ret != IPOURMA_OK)
+		goto free_skb_out;
+	ret = ipourma_update_wr(priv->dev, tx_req);
+	if (unlikely(ret != 0))
+		goto free_skb_out;
+	ret = ubcore_post_jetty_send_wr(priv->jetty[tx_req->eid_index],
+								&tx_req->tx_wr, &jfs_bad_wr);
+	if (unlikely(ret != 0)) {
+		priv->runtime_stats.tx_stats.send_wr_failed++;
+		netdev_dbg(priv->dev, "%s: error code = %d\n",
+					ipourma_err_desc(IPOURMA_POST_SEND_FAILED), ret);
+		goto free_skb_out;
+	}
+	priv->runtime_stats.tx_stats.pass_to_ub++;
+	pr_debug("post_send finish, idx %u, jetty %u\n", tx_req->idx, tx_req->eid_index);
+	return;
+free_skb_out:
+	dev_kfree_skb_any(tx_req->skb);
+	tx_req->skb = NULL;
+}
+
+int ipourma_register_rx_segments(struct net_device *dev,
+	struct ubcore_seg_cfg *cfg, struct ipourma_rx_buf *rx_req)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	u32 blk_idx;
+	u32 i;
+
+	blk_idx = rx_req->idx * priv->skb_buf_size / IPOURMA_REGISTER_SEG_SIZE;
+
+	for (i = 0; i < IPOURMA_MAX_RX_SGES; i++) {
+		rx_req->seg[i] = priv->ipourma_ub_rx_seg[rx_req->eid_index][blk_idx];
+		/* FIME: deal with partial failures */
+		if (IS_ERR_OR_NULL(rx_req->seg[i])) {
+			priv->runtime_stats.rx_stats.register_seg_failed++;
+			netdev_warn(dev, "%s\n",
+						ipourma_err_desc(IPOURMA_REGISTER_SEG_FAILED));
+			return IPOURMA_REGISTER_SEG_FAILED;
+		}
+	}
+
+	return IPOURMA_OK;
 }
 
-void ipourma_replenish_segments(struct work_struct *work)
+STATIC struct sk_buff *ipourma_alloc_rx_skb(struct net_device *dev)
 {
-	/* this function will be filled in the next commit */
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	struct sk_buff *skb;
+
+	skb = alloc_skb(priv->skb_buf_size, GFP_KERNEL);
+	if (IS_ERR_OR_NULL(skb)) {
+		priv->runtime_stats.rx_stats.alloc_skb_failed++;
+		netdev_dbg(dev, "%s\n",
+					ipourma_err_desc(IPOURMA_ALLOC_RX_SKB_FAILED));
+		return NULL;
+	}
+	skb_push(skb, skb_headroom(skb));
+	skb_trim(skb, 0);
+
+	return skb;
 }
 
 STATIC int ipourma_alloc_rx_buffer(struct net_device *dev, u32 eid_idx, u32 idx)
 {
-	/* this function will be filled in the next commit */
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	struct ubcore_seg_cfg cfg = { 0 };
+	struct sk_buff *skb_pass_up;
+	u32 blk_idx, offset;
+
+	skb_pass_up = ipourma_alloc_rx_skb(dev);
+	if (IS_ERR_OR_NULL(skb_pass_up))
+		return IPOURMA_ALLOC_RX_SKB_FAILED;
+
+	blk_idx = idx * priv->skb_buf_size / IPOURMA_REGISTER_SEG_SIZE;
+	offset = (idx * priv->skb_buf_size) % IPOURMA_REGISTER_SEG_SIZE;
+	priv->rx_ring[eid_idx][idx].buf_aligned = priv->rx_buf_aligned[eid_idx][blk_idx] + offset;
+
+	if (ipourma_register_rx_segments(dev, &cfg, &priv->rx_ring[eid_idx][idx])
+			!= IPOURMA_OK) {
+		dev_kfree_skb_any(skb_pass_up);
+		return IPOURMA_REGISTER_SEG_FAILED;
+	}
+
+	priv->rx_ring[eid_idx][idx].skb_pass_up = skb_pass_up;
+	pr_skb_head_plus_linear(skb_pass_up, "Alloced rx buffer skb_pass_up");
+
 	return IPOURMA_OK;
 }
 
 int ipourma_urma_post_recv(struct net_device *dev, u32 eid_idx, u32 idx)
 {
-	/* this function will be filled in the next commit */
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	struct ipourma_rx_buf *rx_buf = &priv->rx_ring[eid_idx][idx];
+	int ret;
+	struct ubcore_jfr_wr *jfr_bad_wr = NULL;
+
+	rx_buf->rx_sge[0].tseg = rx_buf->seg[0];
+	rx_buf->rx_sge[0].addr = (u64)rx_buf->buf_aligned;
+	rx_buf->rx_sge[0].len = priv->urma_mtu;
+	pr_debug("post_recv start, eid idx %u, idx %u\n", rx_buf->eid_index, rx_buf->idx);
+	ret = ubcore_post_jetty_recv_wr(priv->jetty[eid_idx], &rx_buf->rx_wr, &jfr_bad_wr);
+	if (unlikely(ret != 0)) {
+		priv->runtime_stats.rx_stats.post_wr_failed++;
+		netdev_dbg(dev, "%s:%d\n",
+					ipourma_err_desc(IPOURMA_URMA_POST_RECV_FAILED), ret);
+		dev_kfree_skb_any(rx_buf->skb_pass_up);
+		rx_buf->skb_pass_up = NULL;
+		return IPOURMA_URMA_POST_RECV_FAILED;
+	}
+	priv->runtime_stats.rx_stats.num_post_wr++;
+	pr_debug("post_recv finish, eid idx %u, idx %u\n", rx_buf->eid_index, rx_buf->idx);
 	return IPOURMA_OK;
 }
 
@@ -79,6 +406,32 @@ static int ipourma_post_recv_by_eid(struct net_device *dev, u32 eid_idx)
 	return ret;
 }
 
+void ipourma_replenish_segments(struct work_struct *work)
+{
+	struct ipourma_rx_buf *rx_buf = container_of(work, struct ipourma_rx_buf, work);
+	struct ipourma_dev_priv *priv = rx_buf->priv;
+
+	priv->runtime_stats.rx_stats.replenish_deque++;
+
+	/* if post failed, skb will be released as well */
+	rx_buf->skb_pass_up = ipourma_alloc_rx_skb(priv->dev);
+	if (IS_ERR_OR_NULL(rx_buf->skb_pass_up)) {
+		netdev_dbg(priv->dev, "%s: idx = %u\n",
+			ipourma_err_desc(IPOURMA_REPLENISH_RX_SEG_FAILED), rx_buf->idx);
+		return;
+	}
+	if (test_bit(IPOURMA_DEV_ADMIN_UP, &priv->flags))
+		ipourma_urma_post_recv(priv->dev, rx_buf->eid_index, rx_buf->idx);
+}
+
+STATIC inline void ipourma_napi_del(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	netif_napi_del(&priv->napi_send);
+	netif_napi_del(&priv->napi_recv);
+}
+
 int ipourma_urma_init_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx)
 {
 	int ret;
@@ -106,19 +459,149 @@ int ipourma_urma_init_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx)
 
 void ipourma_urma_dev_uninit(struct net_device *dev)
 {
-	/* this function will be filled in the next commit */
+	ipourma_uninit_tjetty_hmap(dev);
+	ipourma_uninit_urma_resources(dev);
+	ipourma_uninit_rings(dev);
+	ipourma_napi_del(dev);
+}
+
+STATIC void ipourma_do_handle_tx_wc(struct net_device *dev,
+	struct ipourma_dev_priv *priv, u32 eid_idx, u32 idx, struct ubcore_cr *cr)
+{
+	struct ipourma_tx_buf *tx_req = &priv->tx_ring[eid_idx][idx];
+
+	if (unlikely(tx_req->tx_buf_in_use == 0))
+		return;
+	tx_req->tx_buf_in_use = 0;
+	ipourma_advance_tx_tail(priv, tx_req);
+	if (unlikely(cr->status != UBCORE_CR_SUCCESS)) {
+		netdev_dbg(dev, "%s: status = %d\n",
+					ipourma_err_desc(IPOURMA_TX_CQE_ERR), cr->status);
+		goto free_tx_skb;
+	}
+	dev->stats.tx_packets++;
+	dev->stats.tx_bytes += tx_req->skb->len;
+	netif_trans_update(dev);
+free_tx_skb:
+	dev_kfree_skb_any(tx_req->skb);
+	tx_req->skb = NULL;
 }
 
 void ipourma_handle_tx_wc(struct net_device *dev,
 			  struct ipourma_dev_priv *priv,
 			  struct ubcore_cr *cr)
 {
-	/* this function will be filled in the next commit */
+	u32 eid_idx, idx;
+
+	if (cr->status >= IPOURMA_MAX_CR_STATUS) {
+		priv->runtime_stats.tx_stats.cqe_err++;
+		return;
+	}
+
+	priv->runtime_stats.tx_stats.cqe_stats[cr->status]++;
+	if (likely(cr->status == UBCORE_CR_SUCCESS))
+		priv->runtime_stats.tx_stats.cqe_success++;
+	else
+		priv->runtime_stats.tx_stats.cqe_err++;
+
+	if (unlikely(cr->status == UBCORE_CR_WR_FLUSH_ERR_DONE)) {
+		/* should not happen, user_ctx is invalid in this case */
+		netdev_dbg(dev, "%s:%d\n",
+				   ipourma_err_desc(IPOURMA_INCORRECT_CR_STATUS), cr->status);
+		return;
+	}
+	if (unlikely(cr->status == UBCORE_CR_FLUSH_ERR))
+		priv->runtime_stats.tx_stats.flush_jetty_success++;
+
+	if (unlikely(cr->local_id < IPOURMA_WELL_KNOWN_JETTY_ID ||
+		cr->local_id >= IPOURMA_MAX_EID_CNT + IPOURMA_WELL_KNOWN_JETTY_ID)) {
+		netdev_dbg(dev, "%s:%u\n",
+				   ipourma_err_desc(IPOURMA_INCORRECT_WQE_JETTY_IDX), eid_idx);
+		return;
+	}
+	eid_idx = cr->local_id - IPOURMA_WELL_KNOWN_JETTY_ID;
+
+	idx = cr->user_ctx;
+	if (unlikely(idx >= ipourma_tx_ring_size)) {
+		netdev_dbg(dev, "%s:%u\n",
+				   ipourma_err_desc(IPOURMA_INCORRECT_WQE_IDX), idx);
+		return;
+	}
+	pr_debug("now cr status:%d\n", cr->status);
+	ipourma_do_handle_tx_wc(dev, priv, eid_idx, idx, cr);
+}
+
+STATIC void ipourma_do_handle_rx_wc(struct net_device *dev,
+					struct ipourma_dev_priv *priv,
+					u32 eid_idx, u32 idx,
+					struct ubcore_cr *cr)
+{
+	struct ipourma_rx_buf *rx_req = &priv->rx_ring[eid_idx][idx];
+	struct sk_buff *skb;
+	u32 data_len;
+
+	/* hold the skb & segments */
+	skb = rx_req->skb_pass_up;
+	rx_req->skb_pass_up = NULL;
+	if (cr->completion_len > priv->tx_buf_size) {
+		dev_kfree_skb_any(skb);
+		priv->runtime_stats.rx_stats.cr_len_err++;
+		goto rx_wc_out;
+	}
+	data_len = cr->completion_len - (u32)sizeof(struct ipourma_header);
+
+	skb->protocol = ((struct ipourma_header *)rx_req->buf_aligned)->proto;
+	memcpy(skb->head, rx_req->buf_aligned + sizeof(struct ipourma_header), data_len);
+	/* update the skb pointer */
+	skb->data_len = 0;
+	skb_trim(skb, 0);
+	skb_put(skb, data_len);
+
+	skb->pkt_type = PACKET_HOST;
+	skb->dev = dev;
+	dev->stats.rx_packets++;
+	dev->stats.rx_bytes += skb->len;
+	pr_debug("IPoURMA received packet's proto: 0x%x\n",
+			 ntohs(skb->protocol));
+	pr_skb_head_plus_linear(skb, "Skb passed to kernel stack");
+	netif_rx(skb);
+	priv->runtime_stats.rx_stats.pass_to_kernel++;
+
+rx_wc_out:
+	priv->runtime_stats.rx_stats.replenish_enque++;
+	queue_work(priv->rx_wq, &(rx_req->work));
 }
 
 void ipourma_handle_rx_wc(struct net_device *dev,
 			  struct ipourma_dev_priv *priv,
 			  struct ubcore_cr *cr)
 {
-	/* this function will be filled in the next commit */
+	u32 eid_idx, idx;
+
+	if (cr->status >= IPOURMA_MAX_CR_STATUS) {
+		priv->runtime_stats.rx_stats.cqe_err++;
+		return;
+	}
+	priv->runtime_stats.rx_stats.cqe_stats[cr->status]++;
+	if (unlikely(cr->status != UBCORE_CR_SUCCESS)) {
+		priv->runtime_stats.rx_stats.cqe_err++;
+		netdev_dbg(dev, "%s:%d\n",
+				   ipourma_err_desc(IPOURMA_INCORRECT_CR_STATUS), cr->status);
+		return;
+	}
+	priv->runtime_stats.rx_stats.cqe_success++;
+	if (unlikely(cr->local_id < IPOURMA_WELL_KNOWN_JETTY_ID ||
+		cr->local_id >= IPOURMA_MAX_EID_CNT + IPOURMA_WELL_KNOWN_JETTY_ID)) {
+		netdev_dbg(dev, "%s:%u\n",
+				   ipourma_err_desc(IPOURMA_INCORRECT_WQE_JETTY_IDX), eid_idx);
+		return;
+	}
+	eid_idx = cr->local_id - IPOURMA_WELL_KNOWN_JETTY_ID;
+	idx = cr->user_ctx;
+	if (unlikely(idx >= ipourma_rx_ring_size)) {
+		netdev_dbg(dev, "%s:%u\n",
+				   ipourma_err_desc(IPOURMA_INCORRECT_WQE_IDX), idx);
+		return;
+	}
+	ipourma_do_handle_rx_wc(dev, priv, eid_idx, idx, cr);
 }
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_ub.h b/drivers/ub/urma/ulp/ipourma/ipourma_ub.h
index cb2b65f05d72..bfc15632e832 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_ub.h
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_ub.h
@@ -12,6 +12,7 @@
 
 void ipourma_build_seg_cfg(struct ubcore_seg_cfg *cfg,
 			   uint64_t va, uint64_t len);
+void ipourma_lru_clear(struct ipourma_tjetty_lru *tjetty_lru);
 void ipourma_init_tjetty_aging_work(struct ipourma_tjetty_lru *tjetty_lru);
 int ipourma_urma_init_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
 void ipourma_urma_dev_uninit(struct net_device *dev);
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_utils.c b/drivers/ub/urma/ulp/ipourma/ipourma_utils.c
new file mode 100644
index 000000000000..0228bc99501f
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_utils.c
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
+ *
+ * Description: ipourma utils
+ */
+#include <linux/kernel.h>
+#include <linux/skbuff.h>
+#include <linux/printk.h>
+#include "ipourma_utils.h"
+
+#define IPOURMA_SKB_HEX_DUMP_ROW_SIZE 16
+#define IPOURMA_SKB_HEX_DUMP_MAX_LEN 128
+
+
+// print first 64 bytes of skb head
+
+void pr_skb_head_plus_linear(const struct sk_buff *skb, const char *prefix)
+{
+	pr_debug("[%s] skb head:0x%llx, linear:0x%llx, total len=%d, ",
+		prefix ? prefix : "IPoURMA", (u64)(skb->head), (u64)(skb->data), skb->len);
+	pr_debug("headroom=%d, linear len=%d, nr_frags=%d\n",
+		skb_headroom(skb), skb_headlen(skb), skb_shinfo(skb)->nr_frags);
+	if ((skb_headroom(skb) + skb_headlen(skb)) > 0) {
+		pr_debug("skb starts from head:\n");
+		print_hex_dump_debug("", DUMP_PREFIX_OFFSET, IPOURMA_SKB_HEX_DUMP_ROW_SIZE,
+					1, skb->head,
+					min_t(size_t, (skb_headroom(skb) + skb_headlen(skb)),
+						IPOURMA_SKB_HEX_DUMP_MAX_LEN),
+					false);
+	}
+}
+
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_utils.h b/drivers/ub/urma/ulp/ipourma/ipourma_utils.h
new file mode 100644
index 000000000000..5d2974fcc5a6
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_utils.h
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2021-2025. All rights reserved.
+ *
+ * Description: Functions definition of ipourma_utils
+ */
+
+#ifndef _IPOURMA_UTILS_H
+#define _IPOURMA_UTILS_H
+
+#include "ipourma_types.h"
+
+void pr_skb_head_plus_linear(const struct sk_buff *skb, const char *prefix);
+
+#endif
-- 
2.43.0

